device/alpaka/src/clusterization/spacepoint_formation_algorithm.cpp

```diff
@@ -0,0 +1,80 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+// Local include(s).
+#include "traccc/alpaka/clusterization/spacepoint_formation_algorithm.hpp"
+
+#include "../utils/barrier.hpp"
+#include "../utils/utils.hpp"
+
+// Project include(s)
+#include "traccc/clusterization/device/form_spacepoints.hpp"
+
+namespace traccc::alpaka {
+
+struct FormSpacepointsKernel {
+    template <typename TAcc>
+    ALPAKA_FN_ACC void operator()(
+        TAcc const& acc,
+        measurement_collection_types::const_view measurements_view,
+        cell_module_collection_types::const_view modules_view,
+        const unsigned int measurement_count,
+        spacepoint_collection_types::view spacepoints_view) const {
+
+        auto const globalThreadIdx =
+            ::alpaka::getIdx<::alpaka::Grid, ::alpaka::Threads>(acc)[0u];
+
+        device::form_spacepoints(globalThreadIdx, measurements_view,
+                                 modules_view, measurement_count,
+                                 spacepoints_view);
+    }
+};
+
+spacepoint_formation_algorithm::spacepoint_formation_algorithm(
+    const traccc::memory_resource& mr, vecmem::copy& copy)
+    : m_mr(mr), m_copy(copy) {}
+
+spacepoint_formation_algorithm::output_type
+spacepoint_formation_algorithm::operator()(
+    const measurement_collection_types::const_view& measurements_view,
+    const cell_module_collection_types::const_view& modules_view) const {
+
+    // Setup alpaka
+    auto devHost = ::alpaka::getDevByIdx(::alpaka::Platform<Host>{}, 0u);
+    auto devAcc = ::alpaka::getDevByIdx(::alpaka::Platform<Acc>{}, 0u);
+    auto queue = Queue{devAcc};
+
+    // Get the number of measurements.
+    const measurement_collection_types::const_view::size_type num_measurements =
+        m_copy.get().get_size(measurements_view);
+
+    // Create the result buffer.
+    spacepoint_collection_types::buffer spacepoints(num_measurements,
+                                                    m_mr.main);
+    m_copy.get().setup(spacepoints)->ignore();
```

Comment:
Does Alpaka use stream ordering? I.e. do we need to wait for this before launching the kernel?

Reply:
Potentially...I don't believe there is any stream ordering (or, at least the vecmem operations aren't part of any alpaka queue at least).